### PR TITLE
DL-14089 Upgrades for HttpVerbs

### PR DESCRIPTION
**Pull request by rajanimohan22**
DL-14089 Upgrades for HttpVerbs

Please include a summary / description of the change and which issue it fixes.  Include any relevant user needs, context or links to other PRs related to this PR (eg. acceptance tests, environment config).

## Checklist Reviewee (add name here)

 - [ ]  I've made every effort to commit high quality, clean code and I have executed relevant static analyses to be sure
 - [ ]  I've included appropriate tests with any code I've added (Unit, Integration, Acceptance etc.)
 - [ ]  I've executed the acceptance test pack locally to ensure there are no functional regressions
 - [ ]  I've added my code using logical, atomic commits, squashing as appropriate - including the JIRA issue number in the commit message
 - [ ]  I've run a dependency check to ensure all dependencies are up to date
 
 ## Checklist Reviewer (add name here)
 
  - [ ]  I've made every effort to commit high quality, clean code and I have executed relevant static analyses to be sure
  - [ ]  I've included appropriate tests with any code I've added (Unit, Integration, Acceptance etc.)
  - [ ]  I've executed the acceptance test pack locally to ensure there are no functional regressions
  - [ ]  I've added my code using logical, atomic commits, squashing as appropriate - including the JIRA issue number in the commit message
  - [ ]  I've run a dependency check to ensure all dependencies are up to date

### CODE REVIEW

File: app/config/BCHandler.scala [hunk 1]
```diff
@@ -16,54 +16,53 @@
...
-                              val templateError: views.html.global_error,
-                              config: ApplicationConfig) extends BCHandler {
-  lazy val appConfig: ApplicationConfig = config
-}
+import java.net.URLEncoder
 
-trait BCHandler extends FrontendErrorHandler with I18nSupport {
-  implicit val appConfig: ApplicationConfig
-  implicit val templateError: views.html.global_error
+class BCHandler @Inject()(val messagesApi: MessagesApi,
```
Comment (cmrhmrc) [inline app/config/BCHandler.scala]: Is this change needed?

File: app/utils/ReferrerUtils.scala [hunk 2]
```diff
@@ -17,13 +17,13 @@
...
-import play.api.mvc.Request
+import play.api.mvc.RequestHeader
 
 object ReferrerUtils {
 
-  def getReferrer()(implicit request: Request[_], appConfig: ApplicationConfig): String =
-    URLEncoder.encode(s"${appConfig.platformHost}${request.path}", "UTF-8")
-
+  def getReferrer()(implicit request: RequestHeader, appConfig: ApplicationConfig): String = {
+    val platFormHost = if(appConfig.platformHost != null) appConfig.platformHost else ""
```
Comment (cmrhmrc) [inline app/utils/ReferrerUtils.scala]: this shouldnt be needed

File: conf/application.conf [hunk 3]
```diff
@@ -18,9 +18,9 @@ appName = "business-customer-frontend"
 
 play.modules.enabled += "config.Wiring"
 
-play.modules.enabled += "uk.gov.hmrc.play.bootstrap.HttpClientModule"
+play.modules.enabled += "uk.gov.hmrc.play.bootstrap.HttpClientV2Module"
 
-play.http.errorHandler = "config.BCHandlerImpl"
+play.http.errorHandler = "config.BCHandler"
```
Comment (cmrhmrc) [inline conf/application.conf]: change related to BCHandler change?

File: app/config/Wiring.scala [hunk 4]
```diff
@@ -31,7 +31,6 @@ class Wiring extends Module {
 
   private def bindControllers: Seq[Binding[_]] = Seq(
     bind[ApplicationConfig].toSelf.eagerly(),
```
Comment (cmrhmrc) [inline app/config/Wiring.scala]: change related to BCHandler change?

File: project/AppDependencies.scala [hunk 5]
```diff
@@ -5,16 +5,15 @@ object AppDependencies {
 
   val compile: Seq[ModuleID] = Seq(
     ws,
-    "uk.gov.hmrc"  %% "bootstrap-frontend-play-30"   % "8.6.0",
-    "uk.gov.hmrc"  %% "domain-play-30"               % "9.0.0",
-    "uk.gov.hmrc"  %% "play-partials-play-30"        % "9.1.0",
-    "uk.gov.hmrc"  %% "play-frontend-hmrc-play-30"   % "10.1.0",
-    "uk.gov.hmrc"  %% "http-caching-client-play-30"  % "11.2.0",
+    "uk.gov.hmrc"  %% "bootstrap-frontend-play-30"   % "9.0.0",
```
Comment (cmrhmrc) [inline project/AppDependencies.scala]: think this si 9.2.0 now

File: test/connectors/BackLinkCacheConnectorSpec.scala [hunk 6]
```diff
@@ -17,44 +17,50 @@
...
       "fetch saved BusinessDetails from SessionCache with Feature Switch on" in new Setup {
         val backLink: BackLinkModel = BackLinkModel(Some("testBackLink"))
 
+
         when(mockSessionCache.fetchAndGetEntry[BackLinkModel](any())(any(), any(), any()))
           .thenReturn(Future.successful(Some(backLink)))
 
-        when(mockHttpClient.GET[CacheMap](any(), any(), any())(any(), any(), any()))
-          .thenReturn(Future.successful(CacheMap("test", Map("BC_Back_Link:testPageId" -> Json.toJson(BackLinkModel(Some("testBackLink")))))))
+        when(mockHttpClient.get(any())(any)).thenReturn(requestBuilder)
```
Comment (cmrhmrc) [inline test/connectors/BackLinkCacheConnectorSpec.scala]: this is already being done in the ConnectorTest

File: test/connectors/BackLinkCacheConnectorSpec.scala [hunk 7]
```diff
@@ -69,9 +75,8 @@ class BackLinkCacheConnectorSpec extends PlaySpec with GuiceOneServerPerSuite wi
         when(mockSessionCache.cache[ReviewDetails](any(), any())(any(), any(), any()))
           .thenReturn(Future.successful(returnedCacheMap))
 
-        when(mockHttpClient.PUT[BackLinkModel, CacheMap]
-          (any(), any(), any())(any(), any(), any(), any()))
-          .thenReturn(Future.successful(CacheMap("test", Map("BC_Back_Link:testPageId" -> Json.toJson(BackLinkModel(Some("testBackLink")))))))
+        when(mockHttpClient.put(any())(any)).thenReturn(requestBuilder)
```
Comment (cmrhmrc) [inline test/connectors/BackLinkCacheConnectorSpec.scala]: not needed, already tsaking place in ConnectorTest class

File: test/connectors/BackLinkCacheConnectorSpec.scala [hunk 8]
```diff
@@ -17,44 +17,50 @@
...
+import uk.gov.hmrc.http.cache.client.SessionCache
+import uk.gov.hmrc.http.cache.client.CacheMap
 
-import scala.concurrent.Future
+import java.util.UUID
+import scala.concurrent.{ExecutionContext, Future}
 
-class BackLinkCacheConnectorSpec extends PlaySpec with GuiceOneServerPerSuite with MockitoSugar with Injecting {
+class BackLinkCacheConnectorSpec extends PlaySpec with GuiceOneServerPerSuite with ConnectorTest with BeforeAndAfterEach {
+
```
Comment (cmrhmrc) [inline test/connectors/BackLinkCacheConnectorSpec.scala]: ConnectorTest should be used on each test rather than the Spec class to ensure no interaction between tests

File: test/connectors/BusinessCustomerConnectorSpec.scala [hunk 9]
```diff
@@ -20,43 +20,42 @@ import audit.Auditable
...
+import uk.gov.hmrc.connectors.ConnectorTest
 import uk.gov.hmrc.http._
 import uk.gov.hmrc.play.audit.DefaultAuditConnector
-import uk.gov.hmrc.play.bootstrap.http.DefaultHttpClient
 
 import java.util.UUID
 import scala.concurrent.{ExecutionContext, Future}
 
-class BusinessCustomerConnectorSpec extends PlaySpec with GuiceOneServerPerSuite with MockitoSugar with BeforeAndAfterEach {
+class BusinessCustomerConnectorSpec extends PlaySpec with GuiceOneServerPerSuite with ConnectorTest with BeforeAndAfterEach {
```
Comment (cmrhmrc) [inline test/connectors/BusinessCustomerConnectorSpec.scala]: ConnectorTest should be used on each test rather than the Spec class to ensure no interaction between tests

File: test/connectors/BusinessCustomerConnectorSpec.scala [hunk 10]
```diff
@@ -71,12 +70,8 @@ class BusinessCustomerConnectorSpec extends PlaySpec with GuiceOneServerPerSuite
         val knownFacts: KnownFactsForService = KnownFactsForService(List(KnownFact("type", "value")))
         val successResponse: JsValue = Json.toJson(knownFacts)
 
-        implicit val hc: HeaderCarrier = HeaderCarrier(sessionId = Some(SessionId(s"session-${UUID.randomUUID}")))
-
-        when(
-          mockHttp.POST[BusinessRegistration, HttpResponse](ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any())
-          (ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any())
-        ).thenReturn(Future.successful(HttpResponse(OK, successResponse.toString)))
+        when(mockHttpClient.post(any())(any)).thenReturn(requestBuilder)
```
Comment (cmrhmrc) [inline test/connectors/BusinessCustomerConnectorSpec.scala]: already done in ConnectorTest

**Review by cmrhmrc — CHANGES_REQUESTED** [review]
The use of ConnectorTest in the connector tests needs to reworked to contruct the class on a test by test basis. Also suggest you replace current ConnectorTest implementation with used within the agent-client-mandate connector tests

File: app/utils/ReferrerUtils.scala (hunk 2 above)
Comment (rajanimohan22) [reply app/utils/ReferrerUtils.scala]: Resolved

File: project/AppDependencies.scala (hunk 5 above)
Comment (rajanimohan22) [reply project/AppDependencies.scala]: Resolved.

File: test/connectors/BackLinkCacheConnectorSpec.scala (hunk 6 above)
Comment (rajanimohan22) [reply test/connectors/BackLinkCacheConnectorSpec.scala]: Resolved.

File: test/connectors/BackLinkCacheConnectorSpec.scala (hunk 7 above)
Comment (rajanimohan22) [reply test/connectors/BackLinkCacheConnectorSpec.scala]: Resolved.

File: test/connectors/BackLinkCacheConnectorSpec.scala (hunk 8 above)
Comment (rajanimohan22) [reply test/connectors/BackLinkCacheConnectorSpec.scala]: Resolved.

File: test/connectors/BusinessCustomerConnectorSpec.scala (hunk 9 above)
Comment (rajanimohan22) [reply test/connectors/BusinessCustomerConnectorSpec.scala]: Resolved.

File: test/connectors/BusinessCustomerConnectorSpec.scala (hunk 10 above)
Comment (rajanimohan22) [reply test/connectors/BusinessCustomerConnectorSpec.scala]: Resolved.

**Review by cmrhmrc — APPROVED** [review]

**Review by nadinJ — APPROVED** [review]
